posts/navarroaxel/09.resolving-git-conflicts-using-a-merge-driver.md

```diff
@@ -0,0 +1,97 @@
+
+I'm sure you know that you can delete the `package-lock.json` file and run `npm install` again - or similar with Yarn.
+
+```bash
+rm package-lock.json
+npm install
+git add package-lock.json
+```
+
+But Git can resolve this automatically (plus this solution works for other auto generated files and binary files, e.g. images).
+
+## The Git attributes
+
+You can create a special file called [`.gitattributes`](https://git-scm.com/docs/gitattributes) placed in the root of your Git repository where you can manually configure how Git handles files. Let's check an example:
+
+```text
+# Auto detect text files and perform LF normalization
+* text=auto
+
+# JS and TS files must always use LF for tools to work
+*.js eol=lf
+*.ts eol=lf
+
+# Use UTF-16 instead of UTF-8 for C source files.
+*.c working-tree-encoding=UTF-16LE-BOM
+
+# Mark all JPEG files as binary.
+*.jpg binary
+
+# Handle as a text file but merge as binary.
+package-lock.json merge=binary
+```
+
+The built-in _binary merge driver_ can't merge binary files, so in case of a conflict Git will leave your `package-lock.json` as-is but marked in a conflicted state.
+
+💡 If you want to keep this attributes for you instead of applying them to every user of a Git repository you can use `.git/info/attributes` file with the same structure.
+
+If you need the changes from the upstream branch just checkout _theirs_ version.
+
+```bash
+git checkout –theirs package-lock.json
+```
+
+This can also be done with a custom merge driver. 👐
+
+## A Git merge driver
+
+A merge driver tries to resolve Git conflicts and you can [create your own](https://git-scm.com/docs/gitattributes#_defining_a_custom_merge_driver) one. Let's define a driver called `theirs` that will use the upstream version (`%B`) over our version (`%A`).
+
+```bash
+git config --global merge.theirs.name "Keep upstream changes"
+git config --global merge.theirs.driver "cp -f '%B' '%A'"
+```
+
+💡 Use `--global` to make this driver available to all repositories in your machine.
+
+Then, add the following to your `.gitattributes` (or `.git/info/attributes`, or `~/.config/git/attributes`) file.
+
+```text
+package-lock.json merge=theirs
+```
+
+Now, every conflict will be resolved by our driver accepting the upstream version. Remember that you should run `npm install` to re-add your changes to the `package-lock.json` because this driver is not smart enough, and discards your changes to the lock file. You need to keep the file in sync with your changes to the `package.json`.
+
+### The npm merge driver
+
+The npm team has created a [npm-merge-drive](https://github.com/npm/npm-merge-driver) which makes a smarter merge than our previous driver - which does an overwrite 😁, keeping the lock file in sync with our changes in the `package.json`. You can just install it globally in your machine without affecting your coworkers because the `~/.config/git/attributes` is used for all your cloned repositories.
```

Comment:
```suggestion
The npm team has created an [npm-merge-drive](https://github.com/npm/npm-merge-driver) which makes a smarter merge than our previous driver - which does an overwrite 😁, keeping the lock file in sync with our changes in the `package.json`. You can just install it globally in your machine without affecting your coworkers because the `~/.config/git/attributes` is used for all your cloned repositories.
```